src/main/java/com/vmware/vipclient/i18n/messages/service/LocaleService.java

```diff
@@ -31,119 +34,166 @@ public LocaleService(LocaleDTO dto) {
-            dispMap = getSupportedDisplayNamesByLocale(fallbackLocale);
-            if (dispMap != null && dispMap.size() > 0) {
-                new FormattingCacheService().addSupportedLanguages(dto, locale, dispMap);
-                logger.debug("Fallback locale [{}] displayNames is cached for product [{}], version [{}], locale [{}]!\n\n",
-                        fallbackLocale, dto.getProductID(), dto.getVersion(), locale);
+            cacheItem = getSupportedDisplayNamesByLocale(fallbackLocale);
+            if (!cacheItem.getCachedData().isEmpty()) {
                 break;
             }
         }
-        return dispMap;
+        return cacheItem.getCachedData();
     }
 
-    public Map<String, String> getSupportedDisplayNamesByLocale(String locale) {
+    public LocaleCacheItem getSupportedDisplayNamesByLocale(String locale) {
         if(locale != null && !locale.isEmpty())
             locale = locale.replace("_", "-").toLowerCase();
-        Map<String, String> dispMap = new HashMap<String, String>();
+        LocaleCacheItem cacheItem = null;
         logger.debug("Look for displayNames from cache for locale [{}]", locale);
         FormattingCacheService formattingCacheService = new FormattingCacheService();
-        dispMap = formattingCacheService.getSupportedLanguages(dto, locale);
-        if (dispMap != null) {
+        cacheItem = formattingCacheService.getSupportedLanguages(dto, locale);
+        if (cacheItem != null) {
+            if (cacheItem.isExpired()) { // cacheItem has expired
+                // Update the cache in a separate thread
+                populateSupportedLanguagesCache(locale, cacheItem);
+            }
             logger.debug("Find displayNames from cache for product [{}], version [{}], locale [{}]!", dto.getProductID(), dto.getVersion(), locale);
-            return dispMap;
+            return cacheItem;
         }
-        dispMap = getSupportedLanguagesFromDS(locale, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
-        if (dispMap != null && dispMap.size() > 0) {
-            logger.debug("Find the displayNames for product [{}], version [{}], locale [{}].\n", dto.getProductID(), dto.getVersion(), locale);
-            formattingCacheService.addSupportedLanguages(dto, locale, dispMap);
-            logger.debug("DisplayNames is cached for product [{}], version [{}], locale [{}]!\n\n", dto.getProductID(), dto.getVersion(), locale);
-            return dispMap;
-        }
-        return dispMap;
+        cacheItem = new LocaleCacheItem();
+        getSupportedLanguagesFromDS(locale, cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
+        formattingCacheService.addSupportedLanguages(dto, locale, cacheItem);
+        logger.debug("DisplayNames is cached for product [{}], version [{}], locale [{}]!\n\n", dto.getProductID(), dto.getVersion(), locale);
```

Comment:
You should not add the cacheItem to the cache fetch if getSupportedLanguagesFromDS failed.
```suggestion
        if (cacheItem.getCachedData().isEmpty() {
            formattingCacheService.addSupportedLanguages(dto, locale, cacheItem);
            logger.debug("DisplayNames is cached for product [{}], version [{}], locale [{}]!\n\n", dto.getProductID(), dto.getVersion(), locale);
        }
```

Reply:
This is for handling non-supported locales, just want to add an empty fetch result to cache, it's the same as you new an empty MessageCacheItem and add it to cache.

Reply:
Same reply:
_If you noticed in my most recent PR, we have changed the design of caching empty cacheItem objects for non-supported locales. This is because this design won't help in the case of the very first request (it will still be slow). So for L3, we just use the list of supported locales to figure out the next steps.. and in the case of non-supported locale, that would be to return fallback locale messages.

For L2, can we do the same in next PR? For now, add the if statement._